src/runtime/hexagon_remote/libadsprpc_shim.cpp

```diff
@@ -0,0 +1,125 @@
+#include <dlfcn.h>
+#include <stdint.h>
+#include <sys/stat.h>
+#include <android/log.h>
+
+// This is a shim to allow dynamic loading of either libcdsprpc.so which
+// supports a compute DSP node on hexagon (SDM660 for example) with HVX.
+// Or an Audio DSP node which supports HVX on 8996/8998 devices (legacy mode).
+
+#define WEAK __attribute__((weak))
+
+namespace {
+
+bool has_cdsp() {
+    const char *cdsp_location = "/dev/subsys_cdsp";
+    struct stat a;
+    if (::stat(cdsp_location, &a) != 0) {
+        return false;
+    }
+    return true;
+}
+
+void *libadsprpc = dlopen(has_cdsp() ? "libcdsprpc.so" : "libadsprpc.so", RTLD_LAZY | RTLD_LOCAL);
```

Comment:
Is it safe to dlopen in a global initializer? I think this is bad on windows with LoadLibrary, but obviously that's different. I might have done that when you took my old branch but that was just a hack.

We're branching in the functions below anyways (checking if the functions are null) so I think we should modify this code like so:

- Initialize this to NULL
- Initialize all the function pointers below to NULL
- In each remote_* function below, if the function pointer is NULL, call get_libadsprpc_symbol
- Make get_libadsprpc_symbol check if libadsprpc is NULL, and if so, then load it

This shouldn't be slower (there's still only one branch taken in the case the function pointers are already loaded), and I think it would be better.

Reply:
Never mind I will go ahead and make these changes

Reply:
Hmm, there are tricky thread safety issues with this, and it *does* work as it is in the PR, maybe we should just leave it alone...

Reply:
i just applied my patch for the CDSP support from our RELEASE_BRANCH to the upstream (halide,llvm) build, rebuilt Halide, and of course rebuilt runtimes using SDK 3.3.0 

i am able to run HelloHexagon on SDM845 in 128 byte mode.

64 byte mode will fail until we get our fix done. 
Pranav is working on it this afternoon

Reply:
Can you clarify, when you say "my patch", are you referring to this branch, or something else?

Also, is rebuilding the runtime with SDK 3.3.0 actually necessary? It seems like the interfaces this runtime relies on should be very stable.

Reply:
a very similar, probably identical patch to what you just merged. I think you have it all working well. Thinking more about the SDK 3.3.0 reference, i think you are correct, you can build with the stable 
version you have used.